### DISCUSSION TITLE
ElasticSearch problem for conductor community server

### DISCUSSION THREAD

**Discussion by spasaria** [Netflix/conductor-community#66] [Q&A]
## Conductor installation process

Good morning, 
I've been trying for a few days to install conductor **community server** trying to interpret the guide, not very exhaustive, of the [readme](https://github.com/Netflix/conductor-community/blob/main/community-server/README.md).

I want to share with you my installation process to see if i did all operations in right way.

### Installation Phases
1. I cloned the two repository: [community](https://github.com/Netflix/conductor-community) and [main repo](https://github.com/Netflix/conductor)

2. In the community folder i ran (at the root folder) the following command to build the server 
                                            `./gradlew build -x test --stacktrace`

3. I copied the community-server folder (just built) to the main repo root folder and I deleted the server folder that existed from the beginning (so I replaced the conductor server with the community server just built).
4. At this point I fixed the docker file of main repo (_conductor/docker/server/Dockerfile_) removing line 17 (the build command) and fixing line 32 for the copy of the jar by the right community server folder (folder that i added manually after the build).
I also fixed the startup.sh file in _conductor/docker/server/bin_ at line 36

At this point i created the image for docker by running the following command:
`docker build -f docker/server/Dockerfile -t conductor:server .`

## Received error

Running the image built following the described process i receive the following **error**:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

An attempt was made to call a method that does not exist. The attempt was made from the following location:

    com.netflix.conductor.es6.dao.index.ElasticSearchRestDAOV6.waitForHealthyCluster(ElasticSearchRestDAOV6.java:336)

The following method did not exist:

    'org.elasticsearch.client.Response org.elasticsearch.client.RestClient.performRequest(java.lang.String, java.lang.String, java.util.Map, org.apache.http.Header[])'

The calling method's class, com.netflix.conductor.es6.dao.index.ElasticSearchRestDAOV6, was loaded from the following location:

    jar:file:/app/libs/conductor-community-server-3.10.0-SNAPSHOT-boot.jar!/BOOT-INF/lib/conductor-es6-persistence-3.9.1.jar!/com/netflix/conductor/es6/dao/index/ElasticSearchRestDAOV6.class

The called method's class, org.elasticsearch.client.RestClient, is available from the following locations:

    jar:file:/app/libs/conductor-community-server-3.10.0-SNAPSHOT-boot.jar!/BOOT-INF/lib/elasticsearch-rest-client-7.15.2.jar!/org/elasticsearch/client/RestClient.class

The called method's class hierarchy was loaded from the following locations:

    org.elasticsearch.client.RestClient: jar:file:/app/libs/conductor-community-server-3.10.0-SNAPSHOT-boot.jar!/BOOT-INF/lib/elasticsearch-rest-client-7.15.2.jar!/


Action:

Correct the classpath of your application so that it contains compatible versions of the classes com.netflix.conductor.es6.dao.index.ElasticSearchRestDAOV6 and org.elasticsearch.client.RestClient

2022-06-30 09:38:25.967 ERROR 6 --- [pool-2-thread-1] c.n.c.m.d.MySQLMetadataDAO               : refresh TaskDefs failed

com.netflix.conductor.core.exception.ApplicationException: BACKEND_ERROR - HikariDataSource HikariDataSource (HikariPool-1) has been closed.
        at com.netflix.conductor.mysql.dao.MySQLBaseDAO.getWithTransaction(MySQLBaseDAO.java:136) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        at com.netflix.conductor.mysql.dao.MySQLBaseDAO.lambda$getWithRetriedTransactions$3(MySQLBaseDAO.java:147) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329) ~[spring-retry-1.3.3.jar!/:?]
        at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209) ~[spring-retry-1.3.3.jar!/:?]
        at com.netflix.conductor.mysql.dao.MySQLBaseDAO.getWithRetriedTransactions(MySQLBaseDAO.java:147) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        at com.netflix.conductor.mysql.dao.MySQLBaseDAO.withTransaction(MySQLBaseDAO.java:194) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        at com.netflix.conductor.mysql.dao.MySQLMetadataDAO.refreshTaskDefs(MySQLMetadataDAO.java:476) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
        at java.util.concurrent.FutureTask.runAndReset(Unknown Source) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
        at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.sql.SQLException: HikariDataSource HikariDataSource (HikariPool-1) has been closed.
        at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:96) ~[HikariCP-4.0.3.jar!/:?]
        at com.netflix.conductor.mysql.dao.MySQLBaseDAO.getWithTransaction(MySQLBaseDAO.java:119) ~[conductor-mysql-persistence-3.10.0-SNAPSHOT.jar!/:3.10.0-SNAPSHOT]
        ... 12 more
```
## Consideration 

As you can see the problem is related to _ElasticSearchRestDAOV6_. 
The ElasticSearch instace that i used in this attempt is 7.6.2 as  indicated in _dependecies.gradle_ file (inside community repo).
I did another attempt using ElasticSearch 7.15.2 because as you can read from log stack error above elasticsearch-rest-client-7.15.2.jar is called.

### Questions 
So, 

- why it try to takes _ElasticSearchRestDAOV6_ ?
- What i have to change?
- I have to add something in config.properties?

The properties used are showed below:

```
    conductor.grpc-server.enabled=false

    conductor.db.type=mysql

    spring.datasource.url=jdbc:mysql://mysql##############/conductor?serverTimezone=UTC

    spring.datasource.username=myUsername

    spring.datasource.password=myPassword

    spring.datasource.hikari.maximum-pool-size=100

    spring.datasource.hikari.minimum-idle=2

    conductor.indexing.enabled=true

    conductor.elasticsearch.url=http://es:9200

    conductor.elasticsearch.indexPrefix=conductor

    loadSample=false
```

**Comment by ghost** [comment]
any further fix?

**Reply by dengguojie** [reply]
same problem . how to fix？？

**Reply by ghost** [reply]
i found a temporary solution, i’m using a maven project to build the server. Since the issue seems to be caused by conflict of version of elasticsearch. so, i use the dependencies related to elasticsearch here, which is version 6, to replace the dependencies in community server. And it solved.

**Reply by dengguojie** [reply]
have you tried to change version in dependencies.lock? @michaelwangroy

**Reply by ghost** [reply]
@dengguojie  same way but not dependencies.lock, i’m not using gradle but maven. so i changed the dependencies in pom file. if you’re going to use elasticsearch 7, you have to change the dependency of elasticsearch in community server to version 7 instead

**Reply by philipsamuel** [reply]
if you are using ES 7 , then use 
conductor.elasticsearch.version=7 instead of 6

**Reply by manan164** [reply]
Hi @dengguojie , If you want to use ES7 then before running,
`./gradlew build -x test --stacktrace`
run,
`./gradlew generateLock updateLock saveLock`
from the root project directory.
This way the dependencies will be updated to use ES7.
Please follow the guide,
`https://github.com/Netflix/conductor-community/tree/main/index`

**Comment by BirknerAlex** [comment]
I had the same issue with running ES7 with the community-server, I figured it out, there are some left-overs in the `community-server/build.gradle`.

This `Dockerfile` should work:

```
FROM openjdk:11-jdk AS builder

ARG RUNTIME_VERSION=v3.13.2

WORKDIR /
RUN git clone -b $RUNTIME_VERSION https://github.com/Netflix/conductor.git
RUN git clone -b $RUNTIME_VERSION https://github.com/Netflix/conductor-community.git

WORKDIR /conductor-community
# Build conductor with all community plugins
RUN find /conductor-community -name dependencies.lock -exec rm -v {} \; && \
    sed -i 's/com.netflix.conductor:conductor-es6-persistence/com.netflix.conductor:conductor-es7-persistence/g' community-server/build.gradle && \
    ./gradlew generateLock updateLock saveLock && \
    ./gradlew build -x test --stacktrace && \
    find /conductor-community/community-server/build/libs/

FROM openjdk:11-jre

RUN mkdir -p \
    /app/config \
    /app/logs \
    /app/libs

COPY --from=builder /conductor/docker/server/bin /app
COPY --from=builder /conductor/docker/server/config /app/config
COPY --from=builder /conductor-community/community-server/build/libs/conductor-community-server-*-boot.jar /app/libs/conductor-server-community-boot.jar

RUN chmod +x /app/startup.sh

CMD [ "/app/startup.sh" ]
ENTRYPOINT [ "/bin/sh"]
```
